scripts/build_test_history.py

```diff
@@ -0,0 +1,447 @@
+        test_name: str,
+        subtest_name: str,
+        run_metadata: MetadataDict,
+        run_date: str,
+        current_status: str) -> TestHistoryEntry:
+    return TestHistoryEntry(
+        RunID=run_metadata['id'],
+        BrowserName=run_metadata['browser_name'],
+        Date=run_date,
+        TestName=test_name,
+        SubtestName=subtest_name,
+        Status=current_status,
+    )
+
+
+def create_entity_if_needed(
+        test_name: str,
+        subtest_name: str,
+        prev_test_statuses: dict,
+        run_metadata: MetadataDict,
+        run_date: str,
+        current_status: str,
+        entities_to_write: list[TestHistoryEntry],
+        unique_entities_to_write: set[tuple[str, str]]) -> None:
+    """Check if an entity should be created for a test status delta,
+    and create one if necessary.
+    """
+    # Test results are stored in dictionary with a tuple key
+    # in the form of (testname, subtest_name).
+    # The overall test status has an empty string as the subtest name.
+    test_key = (test_name, subtest_name)
+    if test_key in unique_entities_to_write:
+        return
+
+    should_create_new_entry = (
+        test_key not in prev_test_statuses or
+        prev_test_statuses[test_key] != current_status)
+
+    if should_create_new_entry:
+        test_status_entry = _build_new_test_history_entry(
+            test_name,
+            subtest_name=subtest_name,
+            run_metadata=run_metadata,
+            run_date=run_date,
+            current_status=current_status
+        )
+        entities_to_write.append(test_status_entry)
+        unique_entities_to_write.add(test_key)
+    prev_test_statuses[test_key] = current_status
+
+
+def process_single_run(run_metadata: MetadataDict) -> None:
+    """Process a single aligned run and save and deltas to history."""
+    try:
+        run_resp = requests.get(run_metadata['raw_results_url'])
+        run_data = run_resp.json()
+    except requests.exceptions.RequestException as e:
+        raise requests.exceptions.RequestException(
+            'Failed to fetch raw results', e)
+
+    # Keep a dictionary of the previous test statuses
+    # from runs we've processed.
+    prev_test_statuses = _populate_previous_statuses(
+        run_metadata['browser_name'])
+
+    # Keep track of every single test result that's in the dataset of
+    # runs we've previously seen. If they're not in the run we're processing,
+    # we'll mark them as missing.
+    tests_not_seen: set[tuple[str, str]] = set(prev_test_statuses.keys())
+
+    run_date = run_metadata["time_start"]
+    entities_to_write: list[TestHistoryEntry] = []
+    unique_entities_to_write: set[tuple[str, str]] = set()
+    # Iterate through each test.
+    for test_data in run_data['results']:
+        # Format the test name.
+        test_name = re.sub(r'\s', ' ', test_data['test'])
+
+        create_entity_if_needed(
+            test_name,
+            subtest_name='',
```

Comment:
```suggestion
            subtest_name='',
```
why do we record a (test_name,  '') pair?

Reply:
An empty string for the subtest name represents the status of the overall test